### ISSUE TITLE
JavaScript object functions (ECMAScript 6)

### ISSUE THREAD

**Issue by marlun** [universal-ctags/ctags#780]
If I've got the following file

```
import Foo from './foo';

const Bar = Foo.create({
  jump() {
    return 123;
  }
})

export default Bar;
```

Bar is found but jump is not. I get the following tags file

```
!_TAG_FILE_FORMAT   2   /extended format; --format=1 will not append ;" to lines/
!_TAG_FILE_SORTED   1   /0=unsorted, 1=sorted, 2=foldcase/
!_TAG_PROGRAM_AUTHOR    Universal Ctags Team    //
!_TAG_PROGRAM_NAME  Universal Ctags /Derived from Exuberant Ctags/
!_TAG_PROGRAM_URL   https://ctags.io/   /official site/
!_TAG_PROGRAM_VERSION   0.0.0   /b6da9a1/
Bar test.js /^const Bar = Foo.create({$/;"  C
```

If I've got the following file

```
import Foo from './foo';

class Bar2 extends Foo {
  jump2() {
    return 123;
  }
}

export default Bar2;
```

Nothing is found where I would like to get Bar2 and jump2. My tags file look like the following

```
!_TAG_FILE_FORMAT   2   /extended format; --format=1 will not append ;" to lines/
!_TAG_FILE_SORTED   1   /0=unsorted, 1=sorted, 2=foldcase/
!_TAG_PROGRAM_AUTHOR    Universal Ctags Team    //
!_TAG_PROGRAM_NAME  Universal Ctags /Derived from Exuberant Ctags/
!_TAG_PROGRAM_URL   https://ctags.io/   /official site/
!_TAG_PROGRAM_VERSION   0.0.0   /b6da9a1/
```

I found the following configuration somewhere on the web which makes ctags find the functions but now it doesn't find Bar, it does however find Bar2.

```
--langdef=js
--langmap=js:.js
--langmap=js:+.jsx

--regex-js=/[ \t.]([A-Z][A-Z0-9._$]+)[ \t]*[=:][ \t]*([0-9"'\[\{]|null)/\1/n,constant/

--regex-js=/\.([A-Za-z0-9._$]+)[ \t]*=[ \t]*\{/\1/o,object/
--regex-js=/['"]*([A-Za-z0-9_$]+)['"]*[ \t]*:[ \t]*\{/\1/o,object/
--regex-js=/([A-Za-z0-9._$]+)\[["']([A-Za-z0-9_$]+)["']\][ \t]*=[ \t]*\{/\1\.\2/o,object/

--regex-js=/([A-Za-z0-9._$]+)[ \t]*=[ \t]*\(function\(\)/\1/c,class/
--regex-js=/['"]*([A-Za-z0-9_$]+)['"]*:[ \t]*\(function\(\)/\1/c,class/
--regex-js=/class[ \t]+([A-Za-z0-9._$]+)[ \t]*/\1/c,class/
--regex-js=/([A-Za-z$][A-Za-z0-9_$()]+)[ \t]*=[ \t]*[Rr]eact.createClass[ \t]*\(/\1/c,class/
--regex-js=/([A-Z][A-Za-z0-9_$]+)[ \t]*=[ \t]*[A-Za-z0-9_$]*[ \t]*[{(]/\1/c,class/
--regex-js=/([A-Z][A-Za-z0-9_$]+)[ \t]*:[ \t]*[A-Za-z0-9_$]*[ \t]*[{(]/\1/c,class/

--regex-js=/([A-Za-z$][A-Za-z0-9_$]+)[ \t]*=[ \t]*function[ \t]*\(/\1/f,function/

--regex-js=/(function)*[ \t]*([A-Za-z$_][A-Za-z0-9_$]+)[ \t]*\([^)]*\)[ \t]*\{/\2/f,function/
--regex-js=/['"]*([A-Za-z$][A-Za-z0-9_$]+)['"]*:[ \t]*function[ \t]*\(/\1/m,method/
--regex-js=/([A-Za-z0-9_$]+)\[["']([A-Za-z0-9_$]+)["']\][ \t]*=[ \t]*function[ \t]*\(/\2/m,method/
```

Any ideas on how to fix this? Is this something that could be changed upstream or is it something that should be added to `~/.ctags` only?

----

( Aug 15 2021 @masatake added more descriptions to make the issue clear.)
The second one is solved in the latest version of u-ctags.
The first one is not.
```
({
  jump() {
    return 123;
  }
})
```
ctags should make a tag for the anonymous class started from `({`.
ctags should make a tag for `jump` method under the scope of the anonymous class.

**Comment by b4n** [comment]
> If I've got the following file
> ```
> import Foo from './foo';
> 
> const Bar = Foo.create({
>   jump() {
>     return 123;
>   }
> })
> 
> export default Bar;
> ```
>
> Bar is found but jump is not.

Not sure if `jump()` here is anything in (pure) JavaScript syntax?  Maybe it's correct ECMAS6, but I don't know this part.

> If I've got the following file
> ```
> import Foo from './foo';
> 
> class Bar2 extends Foo {
>   jump2() {
>     return 123;
>   }
> }
> 
> export default Bar2;
> ```
>
> Nothing is found where I would like to get Bar2 and jump2.

Well, `class` syntax is ECMAS6.  Not supported yet, but could be added (needs proper reference for it, but should be doable).

**Comment by b4n** [comment]
Useful resources
 * http://www.ecma-international.org/publications/files/ECMA-ST/Ecma-262.pdf (esp. §14.5)
 * https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes

**Comment by marlun** [comment]
> Not sure if jump() here is anything in (pure) JavaScript syntax? Maybe it's correct ECMAS6, but I don't know this part.

It is, I know this isn't the prefect reference but at least it describes the features used in my code snippet above:

https://github.com/lukehoban/es6features#enhanced-object-literals

**Comment by masatake** [comment]
The latest ctags works well with the second example input:
```
import Foo from './foo';

class Bar2 extends Foo {
  jump2() {
    return 123;
  }
}

export default Bar2;
```

You will get:
```
Bar2	foo.js	/^class Bar2 extends Foo {$/;"	c
jump2	foo.js	/^  jump2() {$/;"	m	class:Bar2
```

**Comment by masatake** [comment]
The first one, I wonder what kind of tags output we want.
```
const Bar = Foo.create({
  jump() {
    return 123;
  }
})
```
`Bar` should be tagged, and ctags actually makes a tag for it.

`jump` is a method for an anonymous class started from `Foo.create({`?

`node` accepts the following input:
```
({
  jump() {
    return 123;
  }
})
```

ctags don't extract anything from the input.

**Comment by jafl** [comment]
@masatake Please close this, because it was fixed by #3357

**Comment by masatake** [comment]
@jafl, thank you.
